homeassistant/components/thethingsnetwork/entity.py

```diff
@@ -0,0 +1,203 @@
+
+    # ---------------
+    # standard Entity propertiess
+    # ---------------
+
+    @property
+    def unique_id(self) -> str:
+        """Return a unique ID."""
+        return self.get_unique_id(self.device_id, self.field_id)
+
+    @property
+    def name(self) -> Optional[str]:
+        """Return the name of the entity."""
+        return self.__name
+
+    @property
+    def device_info(self) -> DeviceInfo:
+        """Return device specific attributes.
+
+        Implemented by platform classes.
+        """
+
+        return DeviceInfo(
+            {
+                "identifiers": {
+                    # Serial numbers are unique identifiers within a specific domain
+                    (self.__entry.data[CONF_APP_ID], self.device_id)
+                },
+                "name": self.device_name,
+                # TBD - add more info in the TTN upstream message such as signal strength, transmission time, etc
+            }
+        )
+
+    @property
+    def icon(self) -> Optional[str]:
+        """Return the icon to use in the frontend, if any."""
+        return self.__icon
+
+    @property
+    def entity_picture(self) -> Optional[str]:
+        """Return the entity picture to use in the frontend, if any."""
+        return self.__picture
+
+    @property
+    def entity_registry_enabled_default(self) -> bool:
+        """Return if the entity should be enabled when first added to the entity registry."""
+        return True
+
+    # ---------------
+    # TTN integration additional methods
+    # ---------------
+    @property
+    def device_id(self) -> str:
+        """Return device_id."""
+        return str(self._ttn_value.device_id)
+
+    @property
+    def field_id(self) -> str:
+        """Return field_id."""
+        return str(self._ttn_value.field_id)
+
+    @property
+    def device_name(self) -> str:
+        """Return device_name."""
+        return self.__device_name
+
+    @staticmethod
+    @abstractmethod
+    def manages_uplink(
+        entrySettings: TTN_EntrySettings, ttn_value: TTNBaseValue
+    ) -> bool:
+        """Check if this class maps to this ttn_value."""
+
+    def __refresh_names(self) -> None:
+        device_name = self.device_id
+        field_name = self.field_id
+
+        # Device options
+        device_opts = TTN_EntrySettings(self.__entry).get_device_options(self.device_id)
+        device_name = device_opts.get(OPTIONS_DEVICE_NAME, device_name)
+
+        # Field options
+        field_opts = TTN_EntrySettings(self.__entry).get_field_options(
+            self.device_id, self.field_id
+        )
+        field_name = field_opts.get(OPTIONS_FIELD_NAME, field_name)
+        self._unit_of_measurement = field_opts.get(OPTIONS_FIELD_UNIT_MEASUREMENT, None)
+        self.__icon = field_opts.get(OPTIONS_FIELD_ICON, None)
+        self.__picture = field_opts.get(OPTIONS_FIELD_PICTURE, None)
+
+        self.__device_name = device_name
+        self.__name = f"{device_name} {field_name}"
```

Comment:
What do we do here?

Reply:
removed after moving the logic out of the coordinator